src/vumps_localham.jl

```diff
@@ -19,35 +19,63 @@ struct Hᴬᶜ
   n::Int
 end
 
+
+#Assume all local Hamiltonians have the same nrange
 function Base.:*(H::Hᶜ, v::ITensor)
   ∑h = H.∑h
   Hᴸ = H.Hᴸ
   Hᴿ = H.Hᴿ
   ψ = H.ψ
   ψ′ = dag(ψ)'
+  Nsites = nsites(ψ)
+  range_∑h = nrange(∑h, 1)
   n = H.n
   l = linkinds(only, ψ.AL)
   l′ = linkinds(only, ψ′.AL)
   r = linkinds(only, ψ.AR)
   r′ = linkinds(only, ψ′.AR)
+  s = siteinds(only, ψ)
+  s′ = siteinds(only, ψ′)
+  δˢ(n) = δ(s[n], s′[n])
+
   δˡ = δ(l[n], l′[n])
-  δˡ⁻¹ = δ(l[n - 1], l′[n - 1])
   δʳ = δ(r[n], r′[n])
-  δʳ⁺¹ = δ(r[n + 1], r′[n + 1])
   Hᶜᴸv = v * Hᴸ[n] * dag(δʳ)
   Hᶜᴿv = v * δˡ * Hᴿ[n]
-  Hᶜʰv =
-    v * ψ.AL[n] * δˡ⁻¹ * ψ′.AL[n] * ∑h[(n, n + 1)] * ψ.AR[n + 1] * dag(δʳ⁺¹) * ψ′.AR[n + 1]
+  # TODO do we prefer a formalism like the commented one? if yes, need to define contraction of ITensor and MPS
+  #Hᶜʰv = v * ψ.AL[n] * δ(l[n - 1], l′[n - 1]) * ψ′.AL[n] * ∑h[(n, n + 1)] *
+  #  ψ.AR[n+1:n + range_∑h - 1]... * ψ′.AR[n+1:n + range_∑h - 1]... * dag(δ(r[n+range_∑h - 1], r′[n+range_∑h - 1]))
+  Hᶜʰv = v * ψ.AL[n] * δ(l[n - 1], l′[n - 1]) * ψ′.AL[n] * ∑h[(n, n + 1)]
+  for k in 1:range_∑h-2
+      Hᶜʰv =  Hᶜʰv * ψ.AR[n + k] * ψ′.AR[n + k]
+  end
+  Hᶜʰv =  Hᶜʰv * ψ.AR[n + range_∑h - 1] * dag(δ(r[n+range_∑h - 1], r′[n+range_∑h - 1])) * ψ′.AR[n + range_∑h - 1]
+  for j in 1:range_∑h-2
+    temp_Hᶜʰv = ψ.AL[n-j] * δ(l[n - 1 - j], l′[n - 1 - j]) * ψ′.AL[n - j] * ∑h[(n - j, n + 1 - j)];
+    for k in 1:j
+        temp_Hᶜʰv  =  temp_Hᶜʰv  * ψ.AL[n - j + k] * ψ′.AL[n - j + k]
+    end
+    temp_Hᶜʰv = temp_Hᶜʰv  * v
+    for k in j+1:range_∑h-2
+        temp_Hᶜʰv =  temp_Hᶜʰv * ψ.AR[n - j + k] * ψ′.AR[n - j + k]
+    end
+    temp_Hᶜʰv =  temp_Hᶜʰv * ψ.AR[n - j + range_∑h - 1] * dag(δ(r[n - j +  range_∑h - 1], r′[n - j +  range_∑h - 1])) * ψ′.AR[n - j + range_∑h - 1]
+    Hᶜʰv = Hᶜʰv + temp_Hᶜʰv
```

Comment:
I'll have to take your word that this is all correct (very easy to make off-by-one errors in code like this, as I'm sure you know), but it looks reasonable and doesn't seem overly complicated. Could you add a test of this functionality in some way? Even just a full VUMPS run with some longer range interactions where either you know the exact result or reproduce an energy that you think is correct.

Reply:
Also obviously it is a good sign if the 2-local version of VUMPS is not affected by the addition of this code.

Reply:
Ok. We have also seen some convergence issues with gapless models. At larger bond dimensions, we've seen the precision error fluctuating to order 1, which in practice can be improved by using a higher precision in the eigensolver. For example, you can try setting the new keyword argument `eigsolve_tol` introduced in #30 to `eigsolve_tol = (x -> x / 1000)` instead of the default `eigsolve_tol = (x -> x / 100)`. Not sure if that is related to the issue you are seeing though.